model/action.go

```diff
@@ -0,0 +1,1070 @@
+		return action, nil
+	case string(ActionTypeURI):
+		ctx, err := r.newPdfActionURIFromDict(d)
+		if err != nil {
+			return nil, err
+		}
+		ctx.PdfAction = action
+		action.context = ctx
+		return action, nil
+	case string(ActionTypeSound):
+		ctx, err := r.newPdfActionSoundFromDict(d)
+		if err != nil {
+			return nil, err
+		}
+		ctx.PdfAction = action
+		action.context = ctx
+		return action, nil
+	case string(ActionTypeMovie):
+		ctx, err := r.newPdfActionMovieFromDict(d)
+		if err != nil {
+			return nil, err
+		}
+		ctx.PdfAction = action
+		action.context = ctx
+		return action, nil
+	case string(ActionTypeHide):
+		ctx, err := r.newPdfActionHideFromDict(d)
+		if err != nil {
+			return nil, err
+		}
+		ctx.PdfAction = action
+		action.context = ctx
+		return action, nil
+	case string(ActionTypeNamed):
+		ctx, err := r.newPdfActionNamedFromDict(d)
+		if err != nil {
+			return nil, err
+		}
+		ctx.PdfAction = action
+		action.context = ctx
+		return action, nil
+	case string(ActionTypeSubmitForm):
+		ctx, err := r.newPdfActionSubmitFormFromDict(d)
+		if err != nil {
+			return nil, err
+		}
+		ctx.PdfAction = action
+		action.context = ctx
+		return action, nil
+	case string(ActionTypeResetForm):
+		ctx, err := r.newPdfActionResetFormFromDict(d)
+		if err != nil {
+			return nil, err
+		}
+		ctx.PdfAction = action
+		action.context = ctx
+		return action, nil
+	case string(ActionTypeImportData):
+		ctx, err := r.newPdfActionImportDataFromDict(d)
+		if err != nil {
+			return nil, err
+		}
+		ctx.PdfAction = action
+		action.context = ctx
+		return action, nil
+	case string(ActionTypeSetOCGState):
+		ctx, err := r.newPdfActionSetOCGStateFromDict(d)
+		if err != nil {
+			return nil, err
+		}
+		ctx.PdfAction = action
+		action.context = ctx
+		return action, nil
+	case string(ActionTypeRendition):
+		ctx, err := r.newPdfActionRenditionFromDict(d)
+		if err != nil {
+			return nil, err
+		}
+		ctx.PdfAction = action
+		action.context = ctx
+		return action, nil
+	case string(ActionTypeTrans):
+		ctx, err := r.newPdfActionTransFromDict(d)
+		if err != nil {
+			return nil, err
+		}
+		ctx.PdfAction = action
+		action.context = ctx
+		return action, nil
+	case string(ActionTypeGoTo3DView):
+		ctx, err := r.newPdfActionGoTo3DViewFromDict(d)
+		if err != nil {
+			return nil, err
+		}
+		ctx.PdfAction = action
+		action.context = ctx
+		return action, nil
+	case string(ActionTypeJavaScript):
+		ctx, err := r.newPdfActionJavaScriptFromDict(d)
+		if err != nil {
+			return nil, err
+		}
+		ctx.PdfAction = action
+		action.context = ctx
+		return action, nil
+
+	}
+
+	common.Log.Debug("ERROR: Ignoring unknown action: %s", *actionType)
+	return nil, nil
+}
+
+func (r *PdfReader) newPdfActionGotoFromDict(d *core.PdfObjectDictionary) (*PdfActionGoTo, error) {
+	action := PdfActionGoTo{}
+
+	action.D = d.Get("D")
+
+	return &action, nil
+}
+
+func (r *PdfReader) newPdfActionGotoRFromDict(d *core.PdfObjectDictionary) (*PdfActionGoToR, error) {
```

Comment:
Could introduce a function here to load the filespec separately, if one exists, because the filespec loading code does repeat in a lot of the methods. Something like:
```go
func tryLoadFilespec(obj core.PdfObject) (*PdfFilespec, error) {
	if obj == nil {
		return nil, nil
	}

	return NewPdfFilespecFromObj(obj)
}
```

Then you could just do:
```go
func (r *PdfReader) newPdfActionGotoRFromDict(d *core.PdfObjectDictionary) (*PdfActionGoToR, error) {
	filespec, err := tryLoadFilespec(d.Get("F"))
	if err != nil {
		return nil, err
	}

	return &PdfActionGoToR{
		D:         d.Get("D"),
		NewWindow: d.Get("NewWindow"),
		F:         filespec,
	}, nil
}
```

Could do the same for the rest of the `newPdfActionXYZFromDict` methods as well.